### ISSUE TITLE
RFC 8017 compliance

### ISSUE THREAD

**Issue by joechrisellis** [Devolutions/picky-rs#47]
Hi guys!

Just a quick question regarding the [`RSAPrivateKey`](https://docs.rs/picky-asn1-x509/0.1.0/picky_asn1_x509/private_key_info/struct.RSAPrivateKey.html) struct. Definition below for easy reference:

```
pub struct RSAPrivateKey(Asn1SequenceOf<IntegerAsn1>);

impl RSAPrivateKey {
    pub fn modulus(&self) -> &IntegerAsn1 {
        &(self.0).0[1]
    }

    pub fn public_exponent(&self) -> &IntegerAsn1 {
        &(self.0).0[2]
    }

    pub fn private_exponent(&self) -> &IntegerAsn1 {
        &(self.0).0[3]
    }

    pub fn primes(&self) -> &[IntegerAsn1] {
        &(self.0).0[4..]
    }

    pub fn into_public_components(self) -> (IntegerAsn1, IntegerAsn1) {
        let mut iter = (self.0).0.into_iter();
        iter.next().expect("should not panic");
        (
            iter.next().expect("should not panic"),
            iter.next().expect("should not panic"),
        )
    }
}
```

As far as I can tell, this is missing a few fields from [RFC 8017](https://tools.ietf.org/html/rfc8017.html), where RSA private keys are defined like so:

```
RSAPrivateKey ::= SEQUENCE {
    version           Version,
    modulus           INTEGER,  -- n
    publicExponent    INTEGER,  -- e
    privateExponent   INTEGER,  -- d
    prime1            INTEGER,  -- p
    prime2            INTEGER,  -- q
    exponent1         INTEGER,  -- d mod (p-1)
    exponent2         INTEGER,  -- d mod (q-1)
    coefficient       INTEGER,  -- (inverse of q) mod p
    otherPrimeInfos   OtherPrimeInfos OPTIONAL
}
```

Looks like `RSAPrivateKey` treats everything after index 4 as a prime, thereby omitting `exponent1`, `exponent2`, `coefficient`, and `otherPrimeInfos`. Is my interpretation correct? If this is the case, do you think it would be possible to modify this struct to reflect RFC 8017, such as what we have in [this PR](https://github.com/parallaxsecond/rust-psa-crypto/pull/56)? Would be happy to make the change if you give me the go-ahead.

I'm not super experienced with this stuff, so excuse me if I've got something mixed up!

Thanks!
Joe

cc: @hug-dev

**Comment by awakecoding** [comment]
@bcortier-devolutions any thoughts on this?

**Comment by bcortier-devolutions** [comment]
@hug-dev
> > The picky RsaPrivateKey treats index 4 onwards as 'primes', and appears to omit (?) exponent_1, exponent_2, and coefficient,
> 
> I believe that they might have done that because they RsaPrivateKey is just a vector and maybe the fields after `prime_2` are not defined/needed maybe?
> 
> I think it would definitely make sense to post an issue in [`picky`](https://github.com/Devolutions/picky-rs) to ask +1

You are both correct.
Similarly to [RustCrypto/RSA's parse_private_key_pkcs1](https://github.com/RustCrypto/RSA/blob/e8152949f921fdb24efbeb0c27cba5bccfc8d4fc/src/parse.rs#L169) we never really used the others.
I believe we can consider current behavior a bug: calling `primes` and receiving a vector with other values. :thinking: 
I would be glad to accept your PR to improve this! :smile:  

I suggest changing API to add functions such as:
```rust
    pub fn prime1(&self) -> &IntegerAsn1
    pub fn prime2(&self) -> &IntegerAsn1
    pub fn primes(&self) -> &[IntegerAsn1] // the two primes, and not including exponent etc
    pub fn exponent1(&self) -> &IntegerAsn1
    pub fn exponent2(&self) -> &IntegerAsn1
    pub fn exponents(&self) -> &[IntegerAsn1]
    pub fn coefficient(&self) -> &IntegerAsn1
```
You can also make an attempt at supporting the `otherPrimeInfos` if you want to!

Also current solution wrapping an `Asn1SequenceOf<IntegerAsn1>` is done that way because it was easier to return a slice on primes but I don't really like the fact that it's entirely dependent on a "correct key representation": if too many integers are missing in the DER representation, calling `primes` could actually cause a panic. I don't really like when I have to trust user input.
Maybe we should change that to use a struct instead preventing successful de-serialization if anything is missing, that would prevent us from returning a slice on multiple values but we could just make all the fields `pub` allowing easy ownership transfer. What do you think?

**Comment by joechrisellis** [comment]
Hi!

Thanks for the speedy response! Your comments are very helpful -- I'll prepare a PR to fix this. 😄 

> Maybe we should change that to use a struct instead, preventing successful de-serialization if anything is missing, that would prevent us from returning a slice on multiple values but we could just make all the fields pub allowing easy ownership transfer. What do you think?

I think this is a good idea, will do. 👍 

Thanks again!

**Comment by bcortier-devolutions** [comment]
Great, thank you! :smile:

**Comment by Albibek** [comment]
Hi. Just met the same issue just yesterday. Got a working piece of code that passes `openssl rsa -check`. Maybe it would help you with the PR, especially the inversing part.


```
        let key = RSAPrivateKey::new(&mut rng, 2048).unwrap();

        let modulus = IntegerAsn1::from_signed_bytes_be(key.n().to_bytes_be());
        let public_exponent = IntegerAsn1::from_signed_bytes_be(key.e().to_bytes_be());
        let private_exponent = IntegerAsn1::from_signed_bytes_be(key.d().to_bytes_be());

        let primes = key.primes();
        let prime1 = &primes[0];
        let prime2 = &primes[1];

        let exponent1 = key.d().clone() % (prime1 - 1u16);
        let exponent2 = key.d().clone() % (prime2 - 1u16);
        // tried `(1u16 / prime 2) % prime1`, that didn't pass openssl rsa -check
        // so I had to add a num-bigint-dig crate with inverse operation
        use num_bigint_dig::ModInverse;
        let coefficient = prime2.mod_inverse(prime1).unwrap().to_biguint().unwrap();

        let primes = key
            .primes()
            .into_iter()
            .map(|p| IntegerAsn1::from_signed_bytes_be(p.to_bytes_be()))
            .collect::<Vec<_>>();

        let key = Asn1SequenceOf(vec![
            vec![0].into(),
            modulus,
            public_exponent,
            private_exponent,
            primes[0].clone(),
            primes[1].clone(),
            IntegerAsn1::from_signed_bytes_be(exponent1.to_bytes_be()),
            IntegerAsn1::from_signed_bytes_be(exponent2.to_bytes_be()),
            IntegerAsn1::from_signed_bytes_be(coefficient.to_bytes_be()),
        ]);

        return to_vec(&key)
```

**Comment by bcortier-devolutions** [comment]
Addressed by #48 :tada: 
Thank you all!
